## COMPARANDO AS PERFORMANCES

<hr>

## Introdução

<div style=' text-align: justify; text-justify: inter-word;'>
  <br> Um mesmo problema de programação pode ser resolvido por diferentes ferramentas. Nesse sentido, é importante conhecer e entender essas ferramente, visto que isso possibilita compreender em quais contextos elas são mais adequadas e estáveis, a quais limitações elas respondem melhor, com qual delas você tem mais facilidade de trabalhar. Em contextos nos quais o código está relacionado com o um código utilizado por pesquisadores e empresas, é relevante estudar a performance dos algoritmos a fim de poder manter os sistemas estáveis e de entender quais mudanças seriam benéficas para melhorar as respostas dos programa [1].
   <br>Uma forma de estudar algoritmos e compará-los entre si é a análise de desempenho. Com ela, conseguiremos visualizar aqui, por exemplo, a quantidade de tempo que cada algoritmo gasta para resolver o problema das caixas binárias.
</div>

<div style=' text-align: justify; text-justify: inter-word;'>
  <br> No primeiro semestre do curso de Bacharelado em Ciência e Tecnologia, os alunos tem contato com a disciplina de Lógica Computacional. Nela, vimos especificamente opções para se estudar a eficiência de algoritmos. Uma das opções de ferramenta para o estudo foi o módulo 'time', que permite verificar em quanto tempo o computador obtém um resultado e usaremos ela aqui. O Prof. Daniel a definiu da seguinte maneira: 
</div>

 > <center> "Podemos medir o tempo que essa função demora para executar usando o módulo time".</center>

<div style=' text-align: justify; text-justify: inter-word;'>
Além do 'time', já também um outro método conhecido como 'timeit'. Ele também mede o tempo de execução de certo trecho de um código. Todavia, a diferença aqui é que ele roda o código um número definido de vezes e oferece o valor mínimo encontrado entre essas vezes. Para usá-lo, pode-se definir uma função [2], [3].
</div>

## Objetivo

> <div style=' text-align: justify; text-justify: inter-word;'>
   O objetivo do presente experimento consiste em comparar a atividade de três algoritmos distintos no contexto da resolução do problema das caixas binárias. Para tanto, com base nos resultados que obtivermos, buscaremos entender quais deles se destacam e em quais situações. 
</div>

<hr>

## Importações

In [37]:
import random
import time # calcula o tempo que a função gasta 
import timeit # calcula o tempo também, mas com outra metodologia
import numpy as np # análise de caráter estatístico

# para busca aleatória
from funcoes import individuo_cb, funcao_objetivo

# para busca em grade
import itertools 
from funcoes import funcao_objetivo

# para algoritmo genético
from funcoes import populacao_cb as cria_pop_ini 
from funcoes import selecao_roleta_maxima as roleta_max 
from funcoes import funcao_objetivo_populacao_cb as fobj_populacao_cb 
from funcoes import cruzamento_ponto_simples as f_cruzamento
from funcoes import mutacao_cb as f_mutacao

## Códigos e discussão

<hr> 
<center>Todos os códigos apresentados a seguir foram desenvolvidos pelo **Prof. Dr. Daniel Cassar**!</center>
<hr>

> **BUSCA ALEATÓRIA**

In [2]:
# constantes
N = 14 # número de sorteios
N_GENES = 16

iniciando = time.perf_counter()

for n in range(N): 
    candidato = individuo_cb(N_GENES)
    f_objetivo = funcao_objetivo(candidato)
    print(candidato, f_objetivo)
    
finalizando = time.perf_counter()

print()
print('O tempo gasto pelo algoritmo de busca aleatória para a obtenção do resultado foi', finalizando - iniciando, 'segundos.')

[0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0] 5
[1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0] 6
[0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0] 10
[1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0] 10
[0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0] 10
[1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1] 9
[0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0] 9
[1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1] 13
[1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0] 12
[0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0] 7
[0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0] 8
[1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1] 8
[1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1] 9
[0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0] 6

O tempo gasto pelo algoritmo de busca aleatória para a obtenção do resultado foi 0.008988599999999902 segundos.


> **BUSCA EM GRADE**

In [3]:
# primeiro jeito
inicio = time.perf_counter()

for gene_1 in [0, 1]:
    for gene_2 in [0, 1]:
        for gene_3 in [0, 1]:
            for gene_4 in [0, 1]:
                individuo = [gene_1, gene_2, gene_3, gene_4]
                fobj = funcao_objetivo(individuo)
                print(individuo, fobj)

fim = time.perf_counter()

print()
print('O tempo gasto pela primeira opção de busca em grade para a obtenção do resultado foi', fim - inicio, 'segundos.')

[0, 0, 0, 0] 1
[0, 0, 0, 1] 2
[0, 0, 1, 0] 2
[0, 0, 1, 1] 3
[0, 1, 0, 0] 2
[0, 1, 0, 1] 3
[0, 1, 1, 0] 3
[0, 1, 1, 1] 4
[1, 0, 0, 0] 2
[1, 0, 0, 1] 3
[1, 0, 1, 0] 3
[1, 0, 1, 1] 4
[1, 1, 0, 0] 3
[1, 1, 0, 1] 4
[1, 1, 1, 0] 4
[1, 1, 1, 1] 5

O tempo gasto pela primeira opção de busca em grade para a obtenção do resultado foi 0.0019288999999997891 segundos.


In [4]:
# segundo jeito
# esse aqui vai ser mais fácil considerando a mudança de variáveis
inicial = time.perf_counter()

for individuo in itertools.product([0,1], repeat=4):
    fobj = funcao_objetivo(individuo)
    print(individuo, fobj)
    
final = time.perf_counter()

print()
print('O tempo gasto pela segunda opção de busca em grade para a obtenção do resultado foi', final - inicial, 'segundos.')

(0, 0, 0, 0) 1
(0, 0, 0, 1) 2
(0, 0, 1, 0) 2
(0, 0, 1, 1) 3
(0, 1, 0, 0) 2
(0, 1, 0, 1) 3
(0, 1, 1, 0) 3
(0, 1, 1, 1) 4
(1, 0, 0, 0) 2
(1, 0, 0, 1) 3
(1, 0, 1, 0) 3
(1, 0, 1, 1) 4
(1, 1, 0, 0) 3
(1, 1, 0, 1) 4
(1, 1, 1, 0) 4
(1, 1, 1, 1) 5

O tempo gasto pela segunda opção de busca em grade para a obtenção do resultado foi 0.0010298000000004137 segundos.


> **ALGORITMO GENÉTICO**

In [5]:
#constantes
TAMANHO_POP = 6
NUME_GENES = 16
NUME_GERACOES = 57
CHANCE_CRUZAMENTO = 0.5 
CHANCE_MUTACAO = 0.05

inicioo = time.perf_counter() #aplicando o método para a contagem do tempo

populacao = cria_pop_ini(TAMANHO_POP, NUME_GENES) 
print("População Inicial:")
print(populacao)

# tendo a população inicial, agora faremos a seleção e o cruzamento
for n in range(NUME_GERACOES):
    fitness = fobj_populacao_cb(populacao) # seleção dos melhores candidatos de acordo com fitness
    populacao = roleta_max(populacao, fitness) # selecnionando de acordo com os melhores e a proporção na roleta
    pais = populacao[0::2] # corte da população: do elemento 0 até o final, com salto de 2 em dois
    maes = populacao[1::2] # corte da população: do elemento 1 até o final, com intervalo de 2 em 2 
# lembrando que população é uma lista e listas podem ser cortadas
    contador = 0 
    
    for pai, mae in zip(pais, maes):
        if random.random() <= CHANCE_CRUZAMENTO: 
            #vai ter cruzamento
            filho1, filho2 = f_cruzamento(pai, mae)
            populacao[contador] = filho1
            populacao[contador + 1] = filho2
            
        contador = contador + 2
         
    for n in range(len(populacao)):
        if random.random() <= CHANCE_MUTACAO: 
            individuo = populacao[n]
            print()
            print(individuo)
            populacao[n] = f_mutacao(individuo)
            print(populacao[n])
            print()
       
    
print()
print("População Final:")
print(populacao)

finall = time.perf_counter()

print()
print('O tempo gasto pela função para a obtenção do resultado foi', finall - inicioo, 'segundos.')

População Inicial:
[[1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1], [0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0], [0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1], [1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1], [1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0], [1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0]]

[1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]
[1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]


[1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]
[1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]


[1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]
[1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]


[1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]
[1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]


[1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]
[1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1]


[1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]
[1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1]


[1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1]
[1, 0, 0, 1, 0, 0, 0, 1, 1, 1,

> **USANDO O TIMEIT AGORA**

In [6]:
# BUSCA ALEATÓRIA
N = 14 # número de sorteios
NU_GENES = 4

# criando a função para executar o timeit
def busca_aleatoria():
    for n in range(N): 
        candidato = individuo_cb(NU_GENES)
        f_objetivo = funcao_objetivo(candidato)
        #print(candidato, f_objetivo) #rodar isso travou o pc
    
tempo_do_codigo = timeit.timeit(busca_aleatoria, number = 1000) # sendo number o número de excecuções
                                                                # o site mandou 10mil, mas travou o kernel, fazer com 10x menos então
                                                                
print(tempo_do_codigo, 'segundos para achar o resultado da busca aleatória')   

0.046880900000000114 segundos para achar o resultado da busca aleatória


In [23]:
# BUSCA EM GRADE
# sem constantes

def busca_em_grade():
    for gene_1 in [0, 1]:
        for gene_2 in [0, 1]:
            for gene_3 in [0, 1]:
                for gene_4 in [0, 1]:
                    individuo = [gene_1, gene_2, gene_3, gene_4] # precisa reescrever esse código p cada variação
                    fobj = funcao_objetivo(individuo)
        #print(individuo, fobj)
                                                                    
ttempo_codigo = timeit.timeit(busca_em_grade, number = 1000) # sendo number o número de excecuções                                                                
print(ttempo_codigo, 'segundos para achar o resultado da busca aleatória')                                                

0.004189900000028501 segundos para achar o resultado da busca aleatória


In [17]:
# BUSCA EM GRADE
# sem constantes e com itertools

def busca_grade_iter():
    for individuo in itertools.product([0,1], repeat=4):
        fobj = funcao_objetivo(individuo)
        #print(individuo, fobj)
    
t_codigo = timeit.timeit(busca_grade_iter, number = 1000) # sendo number o número de excecuções                                                                
print(t_codigo, 'segundos para achar o resultado da busca aleatória')  

0.007875000000012733 segundos para achar o resultado da busca aleatória


In [10]:
# ALGROTIMO GENÉTICO
# prints comentados apenas para evitar que o computador morra
TAMANHO_POP = 6
NUMERO_GENES = 2
NUMERO_GERACOES = 57
CHANCE_CRUZAMENTO = 0.5 
CHANCE_MUTACAO = 0.05


def busca_alg_genetico():
    populacao = cria_pop_ini(TAMANHO_POP, NUMERO_GENES) 
    #print("População Inicial:")
    #print(populacao)

    for n in range(NUMERO_GERACOES):
        fitness = fobj_populacao_cb(populacao) 
        populacao = roleta_max(populacao, fitness) 
        pais = populacao[0::2] 
        maes = populacao[1::2] 
        contador = 0 
    
    for pai, mae in zip(pais, maes):
        if random.random() <= CHANCE_CRUZAMENTO: 
            filho1, filho2 = f_cruzamento(pai, mae)
            populacao[contador] = filho1
            populacao[contador + 1] = filho2
            
        contador = contador + 2
         
    for n in range(len(populacao)):
        if random.random() <= CHANCE_MUTACAO: 
            individuo = populacao[n]
            #print()
            #print(individuo)
            populacao[n] = f_mutacao(individuo)
            #print(populacao[n])
            #print()
            
tempo_codigo = timeit.timeit(busca_alg_genetico, number = 1000)
print(tempo_codigo, 'segundos para achar o resultado da busca por algoritmo genético')  

0.27878370000007635 segundos para achar o resultado da busca por algoritmo genético


<br>

> **COMPARANDO OS DESEMPENHOS**

**Análise considerando o TEMPO de execução + MUDANÇAS de variáveis**
<div style=' text-align: justify; text-justify: inter-word;'>
    
Para conseguir realizar a análise de performance dos algoritmos de maneira mais robusta, vou não somente **alterar as variáveis, como também alterarei os métodos** e discutirei os resultados para cada um. Sendo assim, vamos começar com o método 'time' acompanhado por mudanças no número de caixas ou, melhor, número de genes. Serão cinco mudanças para cada algoritmo.
<br> Lembrando que os valores de tempo serão alterados conforme eu rodar o código novamente. Nesse sentido, os valores na tabela não são fixos, mas são representativos dos momentos em que eu fazia os testes.
</div>


> Resultados para o algoritmo de BUSCA ALEATÓRIA

| N° caixas | Resultado por TIME (s) | Resultado por TIMEIT (s) |   |   |
|-----------|--------------------|----------------------|---|---|
|     1     |0.0007055000000000256                   |0.029373200000009092  |   |   |
|     4     |0.0005522999999989509                   |0.08292669999991631   |   |   |
|     8     |0.007191799999418436                    |0.14952549999998155   |   |   |
|     12    |0.4781016999999679                      |0.21802279999997154   |   |   |
|     16    |0.00047779999999875145                  |0.2813638999998602    |   |   |

<br> **Breve Análise:**
- Para a maioria dos aumentos, o valor encontrado para o tempo que a função gastou também aumentou. Contudo, nos casos de 4 caixas e 16 caixas, o valor foi inferior, sendo o de 16 caixas o menor de todos.

> <hr>

> Resultados para o algoritmo de BUSCA EM GRADE 

| N° caixas | Resultado por TIME (s) | Resultado por TIMEIT (s) |   |   |
|-----------|--------------------|----------------------|---|---|
|     1     |0.00026270000012118544                  |0.00047270000004573376|   |   |
|     4     |0.0021954999997433333                   |0.006902999999965687  |   |   |
|     8     |0.056430900000123074                    |0.12039559999993799   |   |   |
|     12    |1.3619361000000936                      |1.4276856000001317    |   |   |
|     16    |19.028413                       |23.895921000000044    |   |   |

<br> Uau, 19 segundos me surpreendeu. Achei que o com itertools fosse demorar mais, mas aparentemente ele não é uma ferramenta que agiliza somente o processo de escrever o código, mas facilita a própria operação do computador também.
<br>**Breve Análise:**
- Conforme aumentado o número de caixas, o tempo necessário para obter uma resposta também aumentou.

> ESTATÍSTICA:
<br> Por ser um método determinístico, também é necessário fazer uma análise de cunho mais estatística dos valores de tempo obtidos, a fim do nosso trabalho aqui ser satisfatório. Vejamos ela a seguir:

In [46]:
time_bg = np.array([0.00026270000012118544, 0.0021954999997433333, 0.056430900000123074, 1.3619361000000936, 19.028413])
timeit_bg = np.array([0.00047270000004573376, 0.006902999999965687, 0.12039559999993799, 1.4276856000001317, 23.895921000000044])


print('A média do tempo de execução, usando o método time, para a busca em grade foi de', time_bg.mean(), 'segundos') # média
print('O desvio padrão dos valores referentes ao tempo de execução da busca em grade, usando o método time, é de', time_bg.std(), 'segundos') # desvio padrão

A média do tempo de execução, usando o método time, para a busca em grade foi de 4.089847640000016 segundos
O desvio padrão dos valores referentes ao tempo de execução da busca em grade, usando o método time, é de 7.487379942045018 segundos


> <hr>

> Resultados para o algoritmo de BUSCA EM GRADE COM ITERTOOLS

| N° caixas | Resultado por TIME (s) | Resultado por TIMEIT (s) |   |   |
|-----------|--------------------|----------------------|---|---|
|     1     |0.0002946999999835498                   |0.0007316999999602558 |   |   |
|     4     |0.00033270000005813927                  |0.003851299999951152  |   |   |
|     8     |0.0100524000000064                      |0.09150750000003427   |   |   |
|     12    |0.0008847999997669831                   |1.1771742999999333    |   |   |
|     16    |5.9184338000000025                      |19.505490999999893    |   |   |

<br> O computador gastou demorou tanto para me dar os resultados aqui que eu fiquei com medo de estar fazendo algo errado.
<br> Conversando com outras pessoas que também tentaram fazer esse exercício, eu entendi que talvez tenha dado 'mais trabalho' por conta de ter escolhido  o método de busca em grade que envolve 'itertools'.
<br>**Breve Análise:**
- Aqui, observamos novamente a tendência de aumento do tempo gasto conforme o aumento do número das caixas. Há, porém, uma exceção: o processo para 12 caixas gastou menos tempo.

> ESTATÍSTICA:
<br> Por ser um método determinístico, também é necessário fazer uma análise de cunho mais estatística dos valores de tempo obtidos em grade aletória fazendo o uso de itertools, a fim do nosso trabalho aqui ser satisfatório. Vejamos ela a seguir:

In [47]:
time_bg_iter = np.array([0.0002946999999835498, 0.00033270000005813927, 0.0100524000000064, 0.0008847999997669831, 5.9184338000000025])
timeit_bg_iter = np.array([0.0007316999999602558, 0.003851299999951152, 0.09150750000003427, 1.1771742999999333, 19.505490999999893])

print('A média do tempo de execução, usando o método time, para a busca em grade foi de', time_bg_iter.mean(), 'segundos') # média
print('O desvio padrão dos valores referentes ao tempo de execução da busca em grade, usando o método time, é de', time_bg_iter.std(), 'segundos') # desvio padrão

A média do tempo de execução, usando o método time, para a busca em grade foi de 1.1859996799999635 segundos
O desvio padrão dos valores referentes ao tempo de execução da busca em grade, usando o método time, é de 2.3662199589783097 segundos


> <hr>

> Resultados para o ALGORITMO GENÉTICO

| N° caixas | Resultado por TIME (s) | Resultado por TIMEIT (s) |   |   |
|-----------|--------------------|----------------------|---|---|
|     1     |                    -                   |       -              |   |   |
|     2     |0.0033855000000002633                   |0.5252547999999706    |   |   |
|     4     |0.005741600000021663                    |0.6082541000000106    |   |   |
|     8     |0.004684899999972458                    |0.6014586000000008    |   |   |
|     12    |0.0037179999999921165                   |0.42532620000000065   |   |   |
|     16    |0.0034704999999917163                   |0.49665509999999813   |   |   |

<br>Importante observar: não consegui fazer com uma caixa porque o indivíduo precisa ter mais de um gene para poder dividí-los entre o filho 1 e o filho 2. Testei com o número de genes igual 2, por exemplo, e funcionou.
<br>O número de caixas =2 não entrará na comparação, apenas queria saber quanto daria para um número menor que 4.
<br>**Breve Análise:**
<br>Apesar de ter sim sofrido alterações na performance, essas mudanças foram irrisórias quando comparadas ao demais algoritmos. Entre si, as alterações de tempo também não foram muito significativas apesar das alterações do número de caixas, aqui, genes.
<br>Os dois valores para o maior número de caixas foram os mais inferiores (em timeit) e acredito que isso se deu porque eu rodei o código para eles assim que liguei o comportador, então o Kernel já não estava mais tão sobrecarregado. Vou até fazer um teste para ver se os outros diminuem. Para n =2, por exemplo, o resultado foi 0.27878370000007635. Contudo, como ele foi mudando a cada vez que o código rodar, não vou alterar a planilha.

<hr>

## Conclusão

1. Como um todo, tanto para o método de 'time' quanto para 'timeit', o algoritmo genético apresentou maior estabilidade. O tempo gasto por ele para a obtenção de um resultado foi pequeno em todos os casos e a variação entre eles, apesar do aumento do número das caixas, também não foi gritante. Sendo assim, comparando as performances, a melhor metodologia utilizada é, comprovadamente, o algpritmo genético O que representa o fato de esse recurso ser vantajoso para a resolução de alguns problemas, já que é inclusive mais rápido.
2. Quanto à análise estatística dos métodos determinísticos (busca em grade sem e com itertools), foi possível perceber que a mais vantajosa, tanto nas análises de velocidade quanto no quesito média e desvio padrão, é a busca em grade com itertools.

<HR>

## Referências

[1] SILVESTRE, Iago. **Estudo de Performance de Algoritmos de Controle em Sistema Embarcado**. UFSC, 2020. Disponível em: <https://repositorio.ufsc.br/bitstream/handle/123456789/204775/PFC_Iago_de_Oliveira_Silvestre___modelo_BU_11marc.pdf?sequence=1&isAllowed=y>. 
<br>[2] CAMPBELL, Steve. **Python Timeit() with Examples**. Disponível em: <https://www.guru99.com/timeit-python-examples.html>. 
<br>[3] SIMPLILEARN. **Python timeit()**. 2023. Disponível em: <https://www.simplilearn.com/tutorials/python-tutorial/python-timeit>. 

<hr>

## Playground

# primeiro jeito
inicio = time.perf_counter()

for gene_1 in [0, 1]:
    for gene_2 in [0, 1]:
        for gene_3 in [0, 1]:
            for gene_4 in [0, 1]:
                for gene_5 in [0, 1]:
                    for gene_6 in [0, 1]:
                        for gene_7 in [0, 1]:
                            for gene_8 in [0, 1]:
                                for gene_9 in [0, 1]:
                                    for gene_10 in [0, 1]:
                                        for gene_11 in [0, 1]:
                                            for gene_12 in [0, 1]:
                                                for gene_13 in [0, 1]:
                                                    for gene_14 in [0, 1]:
                                                        for gene_15 in [0, 1]:
                                                            for gene_16 in [0, 1]:
                                                                individuo = [gene_1, gene_2, gene_3, gene_4, gene_5, gene_6, gene_7, gene_8, gene_9, gene_10, gene_11, gene_12, gene_13, gene_14, gene_15, gene_16]
                                                                fobj = funcao_objetivo(individuo)
                                                                print(individuo, fobj)

fim = time.perf_counter()

print()
print('O tempo gasto pela primeira opção de busca em grade para a obtenção do resultado foi', fim - inicio, 'segundos.')